### ISSUE TITLE
missing_panics_doc firing for unreachable situation

### ISSUE THREAD

**Issue by Mythra** [rust-lang/rust-clippy#9564]
### Summary

Clippy seems to be firing `missing_panics_doc` for situations that could never occur. I'm aware clippy seems to in general prefer things like: `if let Some(xxx) = value`, or `match value {`, but in some cases it is just clearer to have an early return/continue/etc. Unfortunately, the missing panics doc lint still fires.

I'm aware this is a lint at the pedantic level, but this is a genuinely useful lint sometimes, especially while proto-typing new code. It allows me to sketch out with things like `unwrap()` handling the happy path, before bubbling up more specific type errors later. The lint helps catch the things i've missed before needing to review the diffs manually. Unfortunately, as more paths have ended up needing to ignore this specific case, I worry the lint will become less useful (as soon a real case might be missed)!

### Lint Name

clippy::missing-panics-doc

### Reproducer

Smallest Code Example I could create:

```rust
pub fn test(aaah: Option<u8>) {
    if aaah.is_none() {
        return;
    }
    let unwrapped = aaah.unwrap();
    println!("{}", unwrapped);
}

pub fn main() {}
```

Then running: `cargo clippy -- -W clippy::pedantic -D warnings`, presents me with:

```txt
error: docs for function which may panic missing `# Panics` section
 --> src/main.rs:1:1
  |
1 | / pub fn test(aaah: Option<u8>) {
2 | |     if aaah.is_none() {
3 | |         return;
4 | |     }
5 | |     let unwrapped = aaah.unwrap();
6 | |     println!("{}", unwrapped);
7 | | }
  | |_^
  |
  = note: `-D clippy::missing-panics-doc` implied by `-D warnings`
note: first possible panic found here
 --> src/main.rs:5:21
  |
5 |     let unwrapped = aaah.unwrap();
  |                     ^^^^^^^^^^^^^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#missing_panics_doc

error: could not compile `tmp` due to previous error
```

### Version

```text
rustc 1.64.0 (a55dd71d5 2022-09-19)
binary: rustc
commit-hash: a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52
commit-date: 2022-09-19
host: x86_64-unknown-linux-gnu
release: 1.64.0
LLVM version: 14.0.6
```

**Comment by llogiq** [comment]
Sorry, but doing control flow analysis is out of scope for the lint, and the idiomatic solution would be to use `let Some(unwrapped) = ahhh else { return; }` (using the recently added let-else). This is also why the lint is in the `pedantic` group which needs a specific opt-in. By doing so, you can expect some warnings that border on false positives.

**Comment by Mythra** [comment]
Thanks, the let some I didn't realize would fix this. However, in hindsight it seems so obvious. 😅 

I figured control-flow was out of the question, but worth a shot right? Thanks for the quick answer!

**Comment by llogiq** [comment]
Perhaps someday we'll have a strong basis on which to easily do this analysis. Until then, we'll pick the low hanging fruit and do our best to help. And in your defence, the `let-else` feature was just introduced with the latest Rust version. Before you'd have to use `if let Some(foo) = bar { .. } else { return; }
